### ISSUE TITLE
Add focus-visible property or ability to hide focus outline when navigating with mouse

### ISSUE THREAD

**Issue by Anastasiia-Boleiko** [Esri/calcite-design-system#8682]
### Check existing issues

- [X] I have [checked for existing issues](https://github.com/Esri/calcite-design-system/issues) to avoid duplicates

### Description

We would like to have an opportunity to enable a **visible outline** on specific components, such as Button and Action, **only when the user interacts with them using the keyboard**. This behavior is desired for our application as many users find the visible outline obtrusive when navigating using a mouse. 
 
We have attempted to achieve this behavior by using the [--calcite-ui-focus-color](https://developers.arcgis.com/calcite-design-system/releases/summaries/2023-05/#focus-color-and-offset-variables) property and combining it with the [:focus-visible](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible). However, we had no luck so far: [Codepen Example](https://codepen.io/olga-knyazeva/pen/eYbNaOG?editors=1100)
 

### Acceptance Criteria

Ability to set:

- [x] `--calcite-ui-focus-color: transparent`
- [ ] `--calcite-ui-focus-visible-color: --calcite-ui-brand`

or 

- [x] `body { --calcite-ui-focus-color: transparent; }`
- [ ] `calcite-button:focus-visible { outline: 2px solid var(--calcite-ui-brand); outline-offset: 2px; }`

### Relevant Info

Additionally, I would like to share a resource that explains the key differences between :focus and :focus-visible . This information might help in understanding the concept better. You can find it here: [Focus vs Focus-visible](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible#focus_vs_focus-visible)
 
Considering the importance of accessibility in the application, we have reviewed the WCAG guidelines and understand that [SC 2.4.7: Focus Visible (Level AA](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html)):

> Any keyboard operable user interface has a mode of operation where the **keyboard focus indicator** is visible.

We have also come across an [article by Deque](https://www.deque.com/blog/give-site-focus-tips-designing-usable-focus-indicators/) that emphasizes the importance of visible focus for elements that can be interacted with using a keyboard.

### Which Component

All interactive components.

### Example Use Case

When navigating with a mouse, after opening a modal window, a close button becomes outlined. Similarly, when clicking on any button, it also has an outline. 
The preferred appearance is to have the outline visible only when the user interacts with these components using the keyboard. However, form components like text fields and selects are still expected to have a visible focus outline, regardless of whether the user navigates with a keyboard or a mouse.

### Priority impact

p2 - want for current milestone

### Calcite package

- [X] @esri/calcite-components
- [ ] @esri/calcite-components-angular
- [X] @esri/calcite-components-react
- [ ] @esri/calcite-design-tokens
- [ ] @esri/eslint-plugin-calcite-components

### Esri team

ArcGIS Urban

**Comment by macandcheese** [comment]
Disabling focus outline for mouse-based navigation can still be a hinderance for low-sighted users. They can also be beneficial to those with attention disorders by providing visual feedback and point-of-reference.

**Comment by macandcheese** [comment]
Disabling focus outline for mouse-based navigation can still be a hinderance for low-sighted users. These focus outlines can also be beneficial to those with attention disorders by providing visual feedback and point-of-reference.

The USWDS is an example of these persistent outlines regardless of interaction device: https://designsystem.digital.gov/components/button/

**Comment by Anastasiia-Boleiko** [comment]
I agree that it could be beneficial for this group of users as well. However, for now, there are a few important reasons why we are looking into this `focus-visible` approach:

 - Users get confused about the focus outline when there is a focus change. For example: 
      - Opening a Calcite Popover with a link automatically outlines the first interactive component that meets accessibility requirements, but users get confused since they haven't clicked on the link and it's not their point-of-reference. <img width="300" alt="Screenshot 2024-02-01 at 21 14 09" src="https://github.com/Esri/calcite-design-system/assets/119309703/ced84ec1-b330-478e-bcc7-5c213e1ab12e">
      - Opening a Calcite Modal outlines the Close button, which also confuses people. <img width="500" alt="Screenshot 2024-02-01 at 21 14 35" src="https://github.com/Esri/calcite-design-system/assets/119309703/1baed3d2-7247-4be3-823e-c19198d2fcfc">
    
 - [Focus-visible is a default](https://developer.mozilla.org/en-US/play) is a default approach in many browsers, so users have become accustomed to it.
 - It's now a WCAG requirement

I believe it would be beneficial for many teams that use the design system to choose a preferred behavior on their side. There are many other important aspects to consider, such as improving accessibility for applications with 2D maps or 3D scenes, where content may not be accessible to keyboard users. However, this is just my personal opinion.

**Comment by macandcheese** [comment]
Seems like this is mostly (only?) causing confusion for users where focus-trapping is involved?

Maybe the focus trap should not focus an element within the trap by default, but on the next tab. This is similar to what USWDC does for Modal - it traps, but doesn't focus immediately: https://designsystem.digital.gov/components/modal/

cc @geospatialem

**Comment by Anastasiia-Boleiko** [comment]
Thank you Adam😊
Sounds like a good alternative and creative approach towards our issue👍

I also noticed something interesting: the focus order in the USWDC Modal allows you to directly proceed to the main/body content of the trapped component. I found it to be a bit more effective.

**Comment by Anastasiia-Boleiko** [comment]
We have received additional feedback stating that the focus outline is not expected when navigating with a mouse. This includes scenarios such as when opening a dropdown and the visual outline moves to the first option, or when clicking a button adds a focus outline. Since there is no evidence that this is a WCAG requirement (either A, AA, or even AAA), I would like to maintain the conversation about the focus-visible property as initially described. The reason for this is that it still aligns with WCAG guidelines and provides a more expected user experience for sighted users. The property will continue to display the focus outline for all interactive components when navigating with a keyboard.

**Comment by geospatialem** [comment]
After some internal accessibility conversations between designers, developers and PE's  Calcite's current offerings are the best way to do so in supporting low vision users who leverage a mouse.

At this time Calcite does not have plans to move in this direction, as the web accessibility community is moving towards support towards low vision mouse user support mentioned in [Success Criterion 2.4.13: Focus Appearance](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance.html) (AAA). Additional documentation and guidance will be made available via https://github.com/Esri/calcite-design-system/issues/8724 to provide more support as to why Calcite's behaves in this way to support more audiences.

**Comment by oevans** [comment]
@macandcheese -- I believe you and Chris H. from the StoryMaps team discussed this recently. We are interested in seeing support for `:focus-visible` to be able to style/manage focus states (one example is web map zoom controls in the JSSDK). Could you consider reopening this issue? I believe other product teams are also asking for this.

**Comment by macandcheese** [comment]
cc @geospatialem for any additional insight /  expertise here - AFAIK this direction was chosen at the recommendation of the a11y team here at Esri.
